sklearn/manifold/t_sne.py

```diff
@@ -674,6 +678,14 @@ def __init__(self, n_components=2, perplexity=30.0,
         self.method = method
         self.angle = angle
 
+        # TODO sklearn 0.20: remove this warning
+        if learning_rate is None:
+            import warnings
+            warnings.warn("The default learning rate of TSNE has changed " +
+                          "from 1000 to 200 in sklearn 0.19. Set the " +
+                          "learning rate explicitely to avoid this message.")
+            self.learning_rate = 200.0
```

Comment:
I doubt `fit` is the right place since it just cals `fit_transform`. Maybe you would like `fit_transform`? But then again, that only calls `_fit`. Maybe `_fit`? Ouch, that only calls `_tsne`. So maybe we got it, it must be in `_tsne`... or not? Because actually `_tsne` only calls `_gradient_descent`....??

Now it *would* work if I changed it in `_gradient_descent`, but `_gradient_descent`:
- is very deep in the hierarchy
- relies heavily on proper instance attributes: it is a numerical recipe, not a playground for object-oriented interfaces

The only way to fix it using "solution 2" while keeping unit tests happy is by breaking even more function signatures and messing around with the deepest numerical function in the code. Or maybe that's going to break more tests? Or maybe some tests in the future? Because we all know burying such fake interfaces deep into the code is bad design.

Bottomline, this argues again for "solution 1", which is just change the constructor default to 200 and tell users about it in the release notes and, if you so wish, upon module import.